pkg/scheduler/objects/application_state.go

```diff
@@ -171,6 +175,9 @@ func NewAppState() *fsm.FSM {
 			},
 			fmt.Sprintf("leave_%s", Running.String()): func(event *fsm.Event) {
 				app := event.Args[0].(*Application) //nolint:errcheck
+				if app.queue != nil {
+					app.queue.runningApps--
```

Comment:
Same here, we need a new `DecRunningApps()`

```
func (sq *Queue) DecRunningApps() {
	sq.Lock()
	defer sq.Unlock()
	sq.runningApps--
	if sq.parent != nil {
		sq.parent.DecRunningApps()
	}
}
```

Reply:
At this point, we have to update the pending applications and start the one which was blocked first.

Just a very POC-like code:
```
pendingApps []*Applications
...
        // TODO: not optimal, a SubmissionTime-based min-Heap is better
	sort.Slice(pendingApps, func(i, j int) bool {
		if pendingApps[i].SubmissionTime.Before(pendingApps[j].SubmissionTime) {
			return true
		}
		return false
	})

	selected := pendingApps[0]
	selected.stateMachine.Event(RunApplication.String(), selected) // trigger app to run
...
```

This is really just a rough idea, need to be tested thoroughly.

Reply:
I think that it just needs to be blocked and doesn't need to start here. It will still be allocated first by the rmEventHandler.

Reply:
Have you tried it?

I think these changes first need to be tested on sample setup with KIND/Minikube to see what happens.

Could you do that? I would like to know what happens.